pkg/controller/route/routecontroller.go

```diff
@@ -33,79 +44,217 @@ type RouteController struct {
+	rc.queue.ShutDown()
+}
+
+func (rc *RouteController) worker() {
+	for {
+		func() {
+			key, quit := rc.queue.Get()
+			if quit {
+				return
+			}
+			defer rc.queue.Done(key)
+			err := rc.syncHandler(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing route: %v", err)
+			}
+		}()
+	}
 }
 
-func (rc *RouteController) reconcileNodeRoutes() error {
-	routeList, err := rc.routes.ListRoutes(rc.clusterName)
+func (rc *RouteController) enqueueNode(obj interface{}) {
+	// node has no namespace, the key of a node is it's name
+	key, err := controller.KeyFunc(obj)
 	if err != nil {
-		return fmt.Errorf("error listing routes: %v", err)
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
 	}
-	// TODO (cjcullen): use pkg/controller/framework.NewInformer to watch this
-	// and reduce the number of lists needed.
-	nodeList, err := rc.kubeClient.Core().Nodes().List(api.ListOptions{})
+	rc.queue.Add(key)
+}
+
+func (rc *RouteController) syncNodeRoute(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing route %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	var node *api.Node
+	obj, exists, err := rc.nodeStore.Store.GetByKey(key)
 	if err != nil {
-		return fmt.Errorf("error listing nodes: %v", err)
+		glog.Infof("Unable to retrieve node %v from store: %v", key, err)
+		rc.queue.Add(key)
+		return err
+	}
+	if exists {
+		node = obj.(*api.Node)
 	}
-	return rc.reconcile(nodeList.Items, routeList)
+
+	var r *cloudprovider.Route
+	routes, err := rc.routes.ListRoutes(rc.clusterName)
+	if err != nil {
+		return fmt.Errorf("error listing routes: %v", err)
+	}
+	// Check if we have a route for this node
+	// TODO(mqliang): we should compute this once and dynamically update it using Watch, not constantly re-compute.
+	for _, route := range routes {
+		if route.TargetInstance != key {
+			continue
+		}
+		r = route
+	}
+
+	var shouldDelete bool
+	var shouldCreate bool
+
+	// if we have a route but the node has been deleted, remove the route
+	if node == nil && r != nil {
+		glog.Infof("Node %v has been deleted, remove it's route", key)
+		shouldDelete = true
+	}
+	// if we have no route for node, create one
+	if node != nil && r == nil {
+		glog.Infof("No route for Node %v, create one", key)
+		shouldCreate = true
+	}
+	// check if the route for this node has the correct CIDR, if not, remove the old route and create a new one
+	if node != nil && r != nil && r.DestinationCIDR != node.Spec.PodCIDR {
+		glog.Infof("The route for %v has the incorrect CIDR, remove it and create a new one", key)
+		shouldDelete = true
+		shouldCreate = true
+	}
+
+	if shouldDelete {
+		if err := rc.routes.DeleteRoute(rc.clusterName, r); err != nil {
+			glog.Errorf("Could not delete route %s %s: %v", r.Name, r.DestinationCIDR, err)
+		}
+	}
+
+	if shouldCreate {
+		route := &cloudprovider.Route{
+			TargetInstance:  node.Name,
+			DestinationCIDR: node.Spec.PodCIDR,
```

Comment:
node.Spec.PodCIDR could still be empty at this point (new nodes), in which case the CreateRoute will fail. We can bail out as soon as we get the node object up above if node.Spec.PodCIDR hasn't been set. Once the cidr allocator has done its thing, we'll get a node update event, and can actually create the route then.

This bug exists in the current code, but since you're here, might as well fix it! :)